Feature/pop up windows

shs69 opened the pull request:
Глобальный контекст сделал через провайдер в App(), если это плохая идея, то сообщи. 
Чуть попозже ещё раз отрефакторю код, уже глаза замылились слегка.

shs69 committed:
Добавлены Popup окна

shs69 committed:
добавлены всплывающие окна уведомлений и добавления контента

shs69 committed:
Финальная версия системы всплывающих окон

Stealr commented on frontend/src/react/app/App.jsx:
для глобальных провайдеров как раз таки и существует файл provider.jsx
перенеси туда PopupProvider

Stealr commented on frontend/src/react/components/cards/CardVideo/CardVideo.jsx:
не очень понимаю зачем нужны были эти изменения
и для нужно два раза повторять нижний двойной прочерк?

shs69 replied:
Скрин с сайта про БЭМ:
<img width="702" alt="image" src="https://github.com/user-attachments/assets/fc9b5fc7-dbcf-4f2d-818c-da0ed278dcbb" />


Я постарался учесть замечания про БЭМ, прочитал после предыдущих задач, мне показалось, что так правильно. Или так наоборот неправильно и я напутал? Я что-то искренне не врубаюсь

Stealr replied:
я просто нигде не видел, чтобы класс состоял из трех и более слов, разделенных ( __ )

у меня следующая логика, каждый компонент начинается с однословного класса(без двойного нижнего подчеркивания, например card-video, тире не считается). Этот класс берется за основу, и следовательно все остальные будут `card-video__<name>`, даже если дальше будет вложенность как с "content".

Допустим если вынести content в отдельный компонент, то компонент должен начинаться с просто "content" и все остальные классы внутри будут `content__<name>`
![image](https://github.com/user-attachments/assets/0d8db069-00eb-4c63-912b-415d080fc023)

хз, написано ли это в доке по БЭМ или это уже моя отсебятина. Но в любом случае придерживаемся данной логики

Stealr commented on frontend/src/react/components/ui/Popaps/Popup.jsx:
```diff
@@ -0,0 +1,82 @@
+            }
+        };
+
+        const handleEsc = (e) => {
+            if (e.key === 'Escape') {
+                onClose();
+            }
+        };
```
за это спасибо, совсем забыл что так надо делать

Stealr commented on frontend/src/react/components/ui/popapContent/notifications/NotificationsContent.jsx:
```diff
@@ -0,0 +1,45 @@
+      </div>
+      <div className="notifications__list title2-small">
+        {mockData.map((item) => (
+          <NotificationItem key={item.id} text={item.text} time={item.time} />
```
стоит делать key уникальным, например вот так: 
``` jsx
key={`notifItem${item.id}`}
```

А то если где ещё будут мок данные с такими же id, то как минимум будет засоряться консоль

Stealr reviewed (changes requested):
### Из технической части
1. Повторный клик на кнопку не закрывает окно

---
### Из визуальной части
1. Нет расстояния между разделяющей линией и первым элементом
![image](https://github.com/user-attachments/assets/f156c49a-c79a-420e-8483-1c967faf0806)

2. расстояние должно быть 4, а не 12 между элементами. Это бросается в глаза
![image](https://github.com/user-attachments/assets/3d70f281-3abb-46d0-990e-9d794a66d64f)

3. Есть ещё части не пиксель перфект, ну да ладно

4. Верхняя часть уезжает наверх. Нужно бы разделить header и блок с уведомлениями
![chrome_AuoWrOG3aT](https://github.com/user-attachments/assets/d632e036-c48d-488d-910a-1ef68c3a8f57)

---
### Просто вопросы
на твой взгляд, возможно ли будет внедрить в данную систему другой контент для popap окна, с другой позицией? Например auth окна
![chrome_Qpl11cPrGm](https://github.com/user-attachments/assets/2d13828c-803a-4790-9521-4de1be1c25c5)

Или лучше создать для этой цели другой контекст?


Также я понял что стоит разбить на компоненты, например avatar. Он везде используется и одинаков. Этим могу я заняться, когда доделаешь popap

shs69 committed:
Исправил замечания:
1. Названия классов привёл к оговорённой логике наименования
2. Провайдер состояния всплывающих окон убрал в Provider, из App.jsx удалил
3. В компоненте аккаунта добавил отступ между "Тема" и белой линией
4. Расстояние в окне "Создать" поменял на 3 пикселя, как в макете (там почему-то не 4, а 3)
5. В окне уведомлений header теперь остаётся на месте, скроллится лишь список уведомлений
6. Ключи в списке уведомлений переделал по твоему примеру
7. Окно теперь закрывается по вторичному нажатию на элемент, который вызвал это окно

shs69 commented:
По поводу переиспользования этой системы для других целей: думаю, что можно. Разве что, я в какой-то степени захардкодил их положение на экране в текущем Popup.jsx, но во-первых положение можно через пропсы прокидывать, во-вторых, если прям совсем плохо будет, то можно закастомить провайдер.
(Например, в PopupRrovider создать поле, которое бы означало, что должно всплыть окно авторизации, кинуть его в value провайдера. Создать callback, который  менял бы это поле, и уже в самом провайдере отображать другой компонент, вместо Popup.jsx)
Наверное не самое лучшее решение, но чем богаты, тем и рады.

Stealr reviewed (approved):
good job

Stealr commented:
> По поводу переиспользования этой системы для других целей: думаю, что можно. Разве что, я в какой-то степени захардкодил их положение на экране в текущем Popup.jsx, но во-первых положение можно через пропсы прокидывать, во-вторых, если прям совсем плохо будет, то можно закастомить провайдер. (Например, в PopupRrovider создать поле, которое бы означало, что должно всплыть окно авторизации, кинуть его в value провайдера. Создать callback, который менял бы это поле, и уже в самом провайдере отображать другой компонент, вместо Popup.jsx) Наверное не самое лучшее решение, но чем богаты, тем и рады.

хорошо, потом посмотрим
